grafana/grafana
# Loki: add from JSON to dataframes tests

(NOTE: this is the same for Loki as https://github.com/grafana/grafana/blob/main/pkg/tsdb/prometheus/framing_test.go is for Prometheus)

there is a small change in `pkg/tsdb/loki/loki.go`, we extract a part of  the `QueryData` function into a separate function, to be able to write tests for it.
the rest of the code is new code, the test code and "input" and "output" snapshot files.

we add tests that test the whole dataframe processing, starting with the Loki JSON response, and seeing how it gets transformed into grafana dataframes.

notes:
1. the tests test what is the current behavior in the Loki alerting code. this may not be what is "correct". (for example, the error-message returned when an error happens is not good)
1. i am testing error-responses too, and those are different in different Loki versions. some versions return it as plaintext, just a text message like "this is wrong", and some other Loki versions return it as a JSON structure. also, some versions return a JSON structure, but use the `text/plain` content type. so i am testing all 3 combinations.


how to test:
- ( this is mostly about adding unit-tests, which are verified by the CI )
- to verify that the backend-changes did not break alerting in Loki, try a simple alerting scenario in Loki, and if it works, it should be ok
